README.md
# rc-week-schedule
---

React SimpleSheet

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][coveralls-image]][coveralls-url]
[![gemnasium deps][gemnasium-image]][gemnasium-url]
[![node version][node-image]][node-url]
[![npm download][download-image]][download-url]

[npm-image]: http://img.shields.io/npm/v/rc-week-schedule.svg?style=flat-square
[npm-url]: http://npmjs.org/package/rc-week-schedule
[travis-image]: https://img.shields.io/travis/pchange/rc-week-schedule.svg?style=flat-square
[travis-url]: https://travis-ci.org/pchange/rc-week-schedule
[coveralls-image]: https://img.shields.io/coveralls/pchange/rc-week-schedule.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/pchange/rc-week-schedule?branch=master
[gemnasium-image]: http://img.shields.io/gemnasium/pchange/rc-week-schedule.svg?style=flat-square
[gemnasium-url]: https://gemnasium.com/pchange/rc-week-schedule
[node-image]: https://img.shields.io/badge/node.js-%3E=_0.10-green.svg?style=flat-square
[node-url]: http://nodejs.org/download/
[download-image]: https://img.shields.io/npm/dm/rc-week-schedule.svg?style=flat-square
[download-url]: https://npmjs.org/package/rc-week-schedule

## Screenshot

<img src="https://raw.githubusercontent.com/pchange/rc-week-schedule/master/screenshot/weekschedule.jpeg" width="450"/>

## Browser Support

|![IE](https://raw.github.com/alrra/browser-logos/master/internet-explorer/internet-explorer_48x48.png) | ![Chrome](https://raw.github.com/alrra/browser-logos/master/chrome/chrome_48x48.png) | ![Firefox](https://raw.github.com/alrra/browser-logos/master/firefox/firefox_48x48.png) | ![Opera](https://raw.github.com/alrra/browser-logos/master/opera/opera_48x48.png) | ![Safari](https://raw.github.com/alrra/browser-logos/master/safari/safari_48x48.png)|
| --- | --- | --- | --- | --- |
| IE 8 + IE 8+ ✔ | Chrome 31.0+ ✔ | Firefox 31.0+ ✔ | Opera 30.0+ ✔ | Safari 7.0+ ✔ |


## Install

[![rc-week-schedule](https://nodei.co/npm/rc-week-schedule.png)](https://npmjs.org/package/rc-week-schedule)

## Usage

```js
var SimpleSheet = require('rc-week-schedule');
var React = require('react');
var ReactDOM = require('react-dom');
ReactDOM.render(<SimpleSheet />, container);
```

## Examples

`npm start` and then go to
[http://localhost:8000/examples](http://localhost:8000/examples)

Online examples: [http://pchange.github.io/rc-week-schedule/examples/](http://pchange.github.io/rc-week-schedule/examples/)

## API

### Props

<table class="table table-bordered table-striped">
    <thead>
    <tr>
        <th style="width: 100px;">name</th>
        <th style="width: 50px;">type</th>
        <th style="width: 50px;">default</th>
        <th>description</th>
    </tr>
    </thead>
    <tbody>
        <tr>
          <td>events</td>
          <td>array</td>
          <td>[]</td>
          <td>components events, if should be as array with elem object link `{start_time: 1499684400, end_time: 1499688000, render: (cell, record) => {return (<span>{record.start_time}</span>); },
        },`</td>
        </tr>
        <tr>
          <td>value</td>
          <td>moment</td>
          <td>momnet()</td>
          <td>components value</td>
        </tr>
        <tr>
          <td>onChange</td>
          <td>Function</td>
          <td></td>
          <td>callback when value changed initiative</td>
        </tr>
        <tr>
          <td>className</td>
          <td>string</td>
          <td>''</td>
          <td>additional className at container</td>
        </tr>
        <tr>
          <td>rowHeight</td>
          <td>number</td>
          <td>60</td>
          <td>change row Height whth this props</td>
        </tr>
        <tr>
          <td>colSpan</td>
          <td>number</td>
          <td>30</td>
          <td>event cell colSpan when rotate</td>
        </tr>
        <tr>
          <td>prefixCls</td>
          <td>string</td>
          <td>rc-week-schedule</td>
          <td>compoments prefixCls class</td>
        </tr>
        <tr>
          <td>direction</td>
          <td>string</td>
          <td>normal</td>
          <td>'rotate' or 'normal'</td>
        </tr>
        <tr>
          <td>showLine</td>
          <td>boolen</td>
          <td>true</td>
          <td>compoments table line segmentation</td>
        </tr>
        <tr>
          <td>header</td>
          <td>function</td>
          <td>`(value) => {
      return (<h3>{value.format('YYYY-MM-DD')}</h3>);
    }`</td>
          <td>compoments header render func</td>
        </tr>
        <tr>
          <td>dayLabelFormat</td>
          <td>function</td>
          <td>`(day) => {
      return <div>{day.format('ddd')}<br />{day.format('MM-DD')}</div>;
    }`</td>
          <td>compoments dayLabelFormat render func</td>
        </tr>
    </tbody>
</table>

## Development

```bash
npm install
npm start
```

## Test Case

```bash
npm test
npm run chrome-test
```

## Coverage

```bash
npm run coverage
```

## License

`rc-week-schedule` is released under the MIT license.


HISTORY.md
# History
---

## 1.0.1 / 2017-09-18

fix readme

## 1.0.0 / 2017-09-18

First release


tests/index.js
import expect from 'expect.js';
import React from 'react';
import { findDOMNode, render, unmountComponentAtNode } from 'react-dom';
// import { Simulate } from 'react-dom/test-utils';
import $ from 'jquery';
import moment from 'moment';
import async from 'async';

import SimpleSheet from '../index';

import '../assets/index.less';

const timeout = (ms) => {
  return (done) => {
    setTimeout(done, ms);
  };
};

const expectQueryLength = (component, query, length) => {
  const componentDomNode = findDOMNode(component);
  const findLength = $(componentDomNode).find(query).length;
  // console.log('query', query);
  expect(findLength).to.be(length);
};

const verifyExist = (component, query, content, done) => {
  async.series([timeout(20), (next) => {
    expectQueryLength(component, query, content);
    next();
  }, timeout(20), (next) => {
    next();
  }], done);
};

const expectPopupToHaveContent = (component, query, content) => {
  const componentDomNode = findDOMNode(component);
  const htmlContent = $(componentDomNode).find(query).text().trim();
  // eslint-disable-next-line
  // console.log('htmlContent', htmlContent, 'content', content, 'query', query);
  expect(htmlContent).to.be(content);
};

const verifyContent = (component, query, content, done) => {
  async.series([timeout(20), (next) => {
    expectPopupToHaveContent(component, query, content);
    next();
  }, timeout(20), (next) => {
    next();
  }], done);
};

describe('rc-week-schedule', () => {
  let div;
  before(() => {
    timeout(40000);
    div = document.createElement('div');
    div.style.margin = '100px';
    document.body.insertBefore(div, document.body.firstChild);
  });

  afterEach(() => {
    unmountComponentAtNode(div);
  });

  describe('check default props', () => {
    it('topLeftCell', (done) => {
      const simpleSheet = render(<SimpleSheet key="default_props_test_1" className="custom-rc-week-schedule" />, div);
      // const prefixCls = simpleSheet.props.prefixCls;
      verifyContent(simpleSheet, 'tr:nth-child(2) td:first-child', '时间', done);
    });
    it('time cell 1', (done) => {
      const simpleSheet = render(<SimpleSheet key="default_props_test_2" className="custom-rc-week-schedule" />, div);
      const prefixCls = simpleSheet.props.prefixCls;
      verifyContent(simpleSheet, `tr:nth-child(3) td:nth-child(1) .${prefixCls}-time-detail:first-child`, '6点', done);
    });
    it('time cell 2', (done) => {
      const simpleSheet = render(<SimpleSheet key="default_props_test_3" className="custom-rc-week-schedule" />, div);
      const prefixCls = simpleSheet.props.prefixCls;
      verifyContent(simpleSheet, `tr:nth-child(3) td:nth-child(1) .${prefixCls}-time-detail:nth-child(2)`, '7点', done);
    });
    it('row cell length default 3', (done) => {
      // this must not exist
      const simpleSheet = render(<SimpleSheet key="default_props_test_4" className="custom-rc-week-schedule" />, div);
      // const prefixCls = simpleSheet.props.prefixCls;
      verifyExist(simpleSheet, 'tr:nth-child(2) td', 8, done);
    });
  });

  describe('check value and events props', () => {
    const props = {
      rowHeight: 40,
      showLine: true,
      direction: 'normal',
      value: moment('2017-07-12'),
      events: [
        {
          start_time: 1499684400,
          end_time: 1499688000,
          name: '小王',
          render: (cell, record) => {
            return (<span>{record.name}</span>);
          },
        },
        {
          start_time: 1499730800,
          end_time: 1499739400,
          name: '小刘',
          render: (cell, record) => {
            return (<span>{record.name}</span>);
          },
        },
        {
          start_time: 1499835400,
          end_time: 1499849500,
          name: '小谢',
          render: (cell, record) => {
            return (<span>{record.name}</span>);
          },
        },
        {
          start_time: 1499930000,
          end_time: 1499937000,
          name: '小航',
          render: (cell, record) => {
            return (<span>{record.name}</span>);
          },
        },
        {
          start_time: 1500025000,
          end_time: 1500030000,
          name: '小张',
          render: (cell, record) => {
            return (<span>{record.name}</span>);
          },
        },
        {
          start_time: 1500035000,
          end_time: 1500082000,
          name: '跨时间的老王',
          render: (cell, record) => {
            return (<span>{record.name}</span>);
          },
        },
      ],
    };
    it('topLeftCell', (done) => {
      const simpleSheet = render(<SimpleSheet {...props} key="default_props_test_1" className="custom-rc-week-schedule" />, div);
      // const prefixCls = simpleSheet.props.prefixCls;
      verifyContent(simpleSheet, 'tr:nth-child(2) td:first-child', '时间', done);
    });
    it('topLeft2Cell', (done) => {
      const simpleSheet = render(<SimpleSheet {...props} key="default_props_test_1" className="custom-rc-week-schedule" />, div);
      // const prefixCls = simpleSheet.props.prefixCls;
      verifyContent(simpleSheet, 'tr:nth-child(2) td:nth-child(2)', 'Mon07-10', done);
    });
    it('topLeft8Cell', (done) => {
      const simpleSheet = render(<SimpleSheet {...props} key="default_props_test_1" className="custom-rc-week-schedule" />, div);
      // const prefixCls = simpleSheet.props.prefixCls;
      verifyContent(simpleSheet, 'tr:nth-child(2) td:nth-child(8)', 'Sun07-16', done);
    });
    it('time detail length', (done) => {
      const simpleSheet = render(<SimpleSheet {...props} key="default_props_test_2" className="custom-rc-week-schedule" />, div);
      const prefixCls = simpleSheet.props.prefixCls;
      verifyExist(simpleSheet, `tr:nth-child(3) td:nth-child(6) .${prefixCls}-cell-column:not(.${prefixCls}-bg-cell-column) .${prefixCls}-time-detail`, 4, done);
    });

    it('time detail cell 1', (done) => {
      const simpleSheet = render(<SimpleSheet {...props} key="default_props_test_2" className="custom-rc-week-schedule" />, div);
      const prefixCls = simpleSheet.props.prefixCls;
      verifyContent(simpleSheet, `tr:nth-child(3) td:nth-child(6) .${prefixCls}-cell-column:not(.${prefixCls}-bg-cell-column) .${prefixCls}-time-detail:nth-child(4)`, '跨时间的老王', done);
    });

    it('time detail cell 2', (done) => {
      const simpleSheet = render(<SimpleSheet {...props} key="default_props_test_2" className="custom-rc-week-schedule" />, div);
      const prefixCls = simpleSheet.props.prefixCls;
      verifyContent(simpleSheet, `tr:nth-child(3) td:nth-child(7) .${prefixCls}-cell-column:not(.${prefixCls}-bg-cell-column) .${prefixCls}-time-detail:nth-child(1)`, '跨时间的老王', done);
    });

    it('time detail cell 3', (done) => {
      const simpleSheet = render(<SimpleSheet {...props} key="default_props_test_2" className="custom-rc-week-schedule" />, div);
      const prefixCls = simpleSheet.props.prefixCls;
      const query = `tr:nth-child(3) td:nth-child(7) .${prefixCls}-cell-column:not(.${prefixCls}-bg-cell-column) .${prefixCls}-time-detail:nth-child(2)`;
      async.series([timeout(20), (next) => {
        const componentDomNode = findDOMNode(simpleSheet);
        const $queryElem = $($(componentDomNode).find(query)[0]);
        expect($queryElem.hasClass('last')).to.be(true);
        expect($queryElem.hasClass(`${prefixCls}-time-detail`)).to.be(true);
        expect($queryElem.hasClass(`${prefixCls}-time-detail-empty`)).to.be(true);
        next();
      }, timeout(20), (next) => {
        next();
      }], done);
    });

    it('time detail cell 4, height', (done) => {
      const simpleSheet = render(<SimpleSheet {...props} key="default_props_test_2" className="custom-rc-week-schedule" />, div);
      const prefixCls = simpleSheet.props.prefixCls;
      const query = `tr:nth-child(3) td:nth-child(7) .${prefixCls}-cell-column:not(.${prefixCls}-bg-cell-column) .${prefixCls}-time-detail:nth-child(1)`;
      async.series([timeout(20), (next) => {
        const componentDomNode = findDOMNode(simpleSheet);
        const $queryElem = $($(componentDomNode).find(query)[0]);
        expect(parseInt($queryElem.css('height'), 10)).to.be(137);
        next();
      }, timeout(20), (next) => {
        next();
      }], done);
    });
    it('time detail cell 5, height', (done) => {
      const simpleSheet = render(<SimpleSheet {...props} key="default_props_test_2" className="custom-rc-week-schedule" />, div);
      const prefixCls = simpleSheet.props.prefixCls;
      const query = `tr:nth-child(3) td:nth-child(7) .${prefixCls}-cell-column:not(.${prefixCls}-bg-cell-column) .${prefixCls}-time-detail:nth-child(2)`;
      async.series([timeout(20), (next) => {
        const componentDomNode = findDOMNode(simpleSheet);
        const $queryElem = $($(componentDomNode).find(query)[0]);
        expect(parseInt($queryElem.css('height'), 10)).to.be(542);
        next();
      }, timeout(20), (next) => {
        next();
      }], done);
    });
  });
});


es/WeekSchedule.js
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { Component } from 'react';
import _ from 'lodash';
import moment from 'moment';
import PropTypes from 'prop-types';

var WeekSchedule = function (_Component) {
  _inherits(WeekSchedule, _Component);

  function WeekSchedule(props) {
    _classCallCheck(this, WeekSchedule);

    var _this = _possibleConstructorReturn(this, (WeekSchedule.__proto__ || Object.getPrototypeOf(WeekSchedule)).call(this, props));

    _this.componentWillMount = function () {};

    _this.componentDidMount = function () {
      _this.countData();
      _this.countMinuteHeight();
      _this.countMinuteColSpan();
    };

    _this.componentWillReceiveProps = function (nextProps) {
      var newState = {};
      if ('value' in nextProps) {
        newState.value = nextProps.value;
      }

      if (!_.isEmpty(newState)) {
        _this.setState(newState, function () {
          _this.countData();
        });
      }
    };

    _this.componentWillUnmount = function () {};

    _this.onChange = function () {
      var onChange = _this.props.onChange;
      var value = _this.state.value;

      if ('function' === typeof onChange) {
        onChange(value);
      }
    };

    _this.makeCellObj = function (_ref) {
      var start = _ref.start,
          end = _ref.end,
          event = _ref.event;

      var length = parseInt((end - start) / 60, 10);
      return {
        start: start,
        end: end,
        length: length,
        event: event,
        empty: !event
      };
    };

    _this.countData = function () {
      var dayLabelFormat = _this.props.dayLabelFormat;
      var format = 'YYYY-MM-DD';
      var repeatValue = moment(_this.state.value.format(format), format);
      var weekFolder = [];

      var _loop = function _loop(i) {
        var day = repeatValue.day(i);
        // 开始时间
        var dayStartUnix = day.unix();
        // 结束时间
        var dayEndUnix = dayStartUnix + 24 * 60 * 60;
        // 筛选所有 events 中该天的 event
        var dayEvents = _.sortBy(_.filter(_this.props.events, function (event) {
          // 这天之前开始，之后结束。
          if (dayStartUnix > event.start_time && dayEndUnix < event.end_time) {
            return true;
          }
          // 这天之前就开始，这天结束。
          else if (dayStartUnix > event.start_time && dayStartUnix < event.end_time) {
              return true;
            }
            // 这天开始，这天结束。
            else if (dayStartUnix < event.start_time && dayEndUnix > event.end_time) {
                return true;
              }
              // 这天开始，这天之后结束。
              else if (dayEndUnix > event.start_time && dayEndUnix < event.end_time) {
                  return true;
                } else {
                  return false;
                }
        }), 'start_time');
        var startMinuteUnix = repeatValue.clone().hour(_this.state.minHour).unix();
        var minuteUnixCount = startMinuteUnix;
        var endMinuteUnix = repeatValue.clone().hour(_this.state.maxHour + _this.state.hourStep).unix();
        var eventsCell = [];
        // 计算每个日期的 event cell 长度。
        _.each(dayEvents, function (event) {
          // 下面两个时间，会通过计算，不能超出当前这天的前后时间。
          var eventStartTime = event.start_time;
          var eventEndTime = event.end_time;

          if (eventStartTime < startMinuteUnix) {
            eventStartTime = startMinuteUnix;
          }
          if (eventStartTime < minuteUnixCount) {
            eventStartTime = minuteUnixCount;
          }
          if (eventEndTime > endMinuteUnix) {
            eventEndTime = endMinuteUnix;
          }
          if (eventEndTime < minuteUnixCount) {
            eventEndTime = minuteUnixCount;
          }

          if (minuteUnixCount < eventStartTime) {
            var emtryEventCell = _this.makeCellObj({
              start: minuteUnixCount,
              end: eventStartTime
            });
            minuteUnixCount = endMinuteUnix;
            eventsCell.push(emtryEventCell);
            minuteUnixCount = eventStartTime;
          }

          var eventCell = _this.makeCellObj({
            start: minuteUnixCount,
            end: eventEndTime,
            event: event
          });
          minuteUnixCount = eventEndTime;
          eventsCell.push(eventCell);
        });
        // 尾部如果空白，需要追加填充一个空白的事件。
        if (minuteUnixCount < endMinuteUnix) {
          // 尾部填充一个空白的日期。
          eventsCell.push(_this.makeCellObj({
            start: minuteUnixCount,
            end: endMinuteUnix
          }));
          minuteUnixCount = endMinuteUnix;
        }

        weekFolder.push({
          day: i,
          date: day.format('MM-DD'),
          dayStartUnix: dayStartUnix,
          dayEndUnix: dayEndUnix,
          events: dayEvents,
          eventsCell: eventsCell,
          label: dayLabelFormat(day)
        });
      };

      for (var i = _this.state.weekStart; i <= _this.state.weekEnd; i += 1) {
        _loop(i);
      }

      _this.setState({
        weekFolder: weekFolder
      });
    };

    _this.countMinuteHeight = function () {
      // 每分钟的高度，这个根据 rowHeight ， hourStep 计算出来，计算公式是
      var minuteHeight = _this.state.rowHeight / (60 * _this.state.hourStep);
      _this.setState({
        minuteHeight: minuteHeight
      });
    };

    _this.countMinuteColSpan = function () {
      // 每分钟的高度，这个根据 colSpan ， 每 row 分钟数计算出来
      var minuteColSpan = _this.state.colSpan / (60 * _this.state.hourStep);
      _this.setState({
        minuteColSpan: minuteColSpan
      });
    };

    _this.renderCell = function (record) {
      if (record.empty) {
        return null;
      } else if (record && record.event && record.event.render) {
        return record.event.render(record, record.event);
      }
    };

    _this.state = {
      // 方向，默认 normal
      // direction: props.direction,
      // 日历当前的值，显示这周的日期。
      value: props.value,
      // 每行的高度
      rowHeight: props.rowHeight,
      // 存储这一周每一天的数据，详情看看内存即可。
      weekFolder: [],
      // 开始显示时间。
      minHour: 6,
      // 结束显示时间。
      maxHour: 22,
      // 每个时间的间隔。
      hourStep: 1,
      // 一周开始的时间
      weekStart: 1,
      // 一周结束的时间
      weekEnd: 7,
      minuteHeight: 1,
      colSpan: props.colSpan
    };
    // 每分钟的高度，这个根据 rowHeight ， hourStep 计算出来，计算公式是
    _this.state.minuteHeight = _this.state.rowHeight / (60 * _this.state.hourStep);
    // 每分钟的 colSpan ，这个根据 rowHeight ， hourStep 计算出来，计算公式是
    _this.state.minuteColSpan = _this.state.colSpan / (60 * _this.state.hourStep);
    return _this;
  }

  // 更新传输的 value


  _createClass(WeekSchedule, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var props = this.props;
      var prefixCls = props.prefixCls;
      var state = this.state;
      var timeLabelArr = _.range(state.minHour, state.maxHour + state.hourStep, state.hourStep);

      return React.createElement(
        'div',
        { className: prefixCls + ' ' + prefixCls + '-direction-' + this.props.direction },
        React.createElement('style', null),
        'rotate' === this.props.direction ? React.createElement(
          'table',
          { cellPadding: '0', cellSpacing: '0', className: prefixCls + '-table print-table' },
          React.createElement(
            'tbody',
            { 'data-comment': '\u6CE8\u610F\u4E0D\u5141\u8BB8\u4F7F\u7528 thead' },
            'function' !== typeof this.props.header ? null : React.createElement(
              'tr',
              null,
              React.createElement(
                'td',
                { colSpan: this.state.colSpan * (1 + timeLabelArr.length) },
                this.props.header(state.value)
              )
            ),
            React.createElement(
              'tr',
              null,
              React.createElement(
                'td',
                { colSpan: this.state.colSpan },
                '\u65F6\u95F4'
              ),
              timeLabelArr.map(function (hour, hourIndex, hourArr) {
                var last = hourIndex === hourArr.length - 1;
                return React.createElement(
                  'td',
                  { colSpan: _this2.state.colSpan, key: 'hourIndex_' + hourIndex, style: { height: state.rowHeight }, className: prefixCls + '-time-detail ' + (last ? 'last' : '') },
                  hour,
                  '\u70B9'
                );
              })
            ),
            state.weekFolder.map(function (day) {
              var eventsCell = day.eventsCell;
              return React.createElement(
                'tr',
                { key: day.date },
                React.createElement(
                  'td',
                  { colSpan: _this2.state.colSpan },
                  day.label
                ),
                eventsCell.map(function (cell, cellIndex) {
                  var last = cellIndex === eventsCell.length - 1;
                  var colSpan = cell.length * state.minuteColSpan;
                  return React.createElement(
                    'td',
                    { key: 'weekIndex_' + cellIndex, colSpan: colSpan },
                    React.createElement(
                      'div',
                      { className: prefixCls + '-time-detail ' + (last ? 'last' : '') + ' ' + (cell.empty ? '${prefixCls}-time-detail-empty' : '${prefixCls}-time-detail-full') },
                      _this2.renderCell(cell)
                    )
                  );
                })
              );
            })
          )
        ) : null,
        'rotate' !== this.props.direction ? React.createElement(
          'table',
          { cellPadding: '0', cellSpacing: '0', className: prefixCls + '-table print-table' },
          React.createElement(
            'tbody',
            { 'data-comment': '\u6CE8\u610F\u4E0D\u5141\u8BB8\u4F7F\u7528 thead' },
            'function' !== typeof this.props.header ? null : React.createElement(
              'tr',
              null,
              React.createElement(
                'td',
                { colSpan: 1 + state.weekFolder.length },
                this.props.header(state.value)
              )
            ),
            React.createElement(
              'tr',
              null,
              React.createElement(
                'td',
                { colSpan: '1' },
                '\u65F6\u95F4'
              ),
              state.weekFolder.map(function (day, index) {
                return React.createElement(
                  'td',
                  { colSpan: '1', key: index },
                  day.label
                );
              })
            ),
            React.createElement(
              'tr',
              null,
              React.createElement(
                'td',
                { colSpan: '1', className: prefixCls + '-cell-td' },
                timeLabelArr.map(function (hour, hourIndex, hourArr) {
                  var last = hourIndex === hourArr.length - 1;
                  return React.createElement(
                    'div',
                    { key: 'hourIndex_' + hourIndex, style: { height: state.rowHeight }, className: prefixCls + '-time-detail ' + (last ? 'last' : '') },
                    hour,
                    '\u70B9'
                  );
                })
              ),
              state.weekFolder.map(function (day) {
                var eventsCell = day.eventsCell;
                return React.createElement(
                  'td',
                  { colSpan: '1', key: day.date, className: prefixCls + '-cell-td' },
                  !_this2.props.showLine ? null : React.createElement(
                    'div',
                    { className: prefixCls + '-cell-column ' + prefixCls + '-bg-cell-column' },
                    timeLabelArr.map(function (hour, hourIndex, hourArr) {
                      var last = hourIndex === hourArr.length - 1;
                      return React.createElement('div', { key: 'hourIndex_' + hourIndex, style: { height: state.rowHeight }, className: prefixCls + '-time-detail ' + (last ? 'last' : '') });
                    })
                  ),
                  React.createElement(
                    'div',
                    { className: prefixCls + '-cell-column' },
                    eventsCell.map(function (cell, cellIndex) {
                      var last = cellIndex === eventsCell.length - 1;
                      var height = cell.length * state.minuteHeight;
                      // eslint-disable-next-line
                      var className = prefixCls + '-time-detail ' + (last ? 'last ' : ' ') + (cell.empty ? prefixCls + '-time-detail-empty ' : prefixCls + '-time-detail-full ');
                      return React.createElement(
                        'div',
                        { key: 'weekIndex_' + cellIndex, style: { height: height }, className: className },
                        _this2.renderCell(cell)
                      );
                    })
                  )
                );
              })
            )
          )
        ) : null
      );
    }
  }]);

  return WeekSchedule;
}(Component);

WeekSchedule.propTypes = {
  // 方向，只有两个，为 normal 和 rotate
  direction: PropTypes.string,
  // 时间表的 value 值。
  value: PropTypes.array,
  // 正常方向时候，每一行高度。
  rowHeight: PropTypes.number,
  // 翻转时候的每个 cell 的 colSpan 。
  colSpan: PropTypes.number,
  // value 变化之后的回调。
  onChange: PropTypes.func,
  // 时间表抬头的 label 的 format 方法。
  dayLabelFormat: PropTypes.func,
  // 时间表的事件。
  events: PropTypes.array,
  // 渲染 header
  header: PropTypes.func,
  // 是否显示表格里面的时间分割线。
  showLine: PropTypes.bool
};
WeekSchedule.defaultProps = {
  prefixCls: 'rc-week-schedule',
  value: moment(),
  events: [],
  showLine: true,
  // direction: 'rotate',
  direction: 'normal',
  rowHeight: 60,
  colSpan: 30,
  header: function header(value) {
    return React.createElement(
      'h3',
      null,
      value.format('YYYY-MM-DD')
    );
  },
  dayLabelFormat: function dayLabelFormat(day) {
    return React.createElement(
      'div',
      null,
      day.format('ddd'),
      React.createElement('br', null),
      day.format('MM-DD')
    );
  }
};


export default WeekSchedule;

examples/simple.js
import React, { Component } from 'react';
import moment from 'moment';
import ReactDOM from 'react-dom';
import WeekSchedule from 'rc-week-schedule';
import 'rc-week-schedule/assets/index.less';

class Test extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: moment('2017-07-12'),
      events: [
        {
          start_time: 1499684400,
          end_time: 1499688000,
          name: '小王',
          render: (cell, record) => {
            return (<span>{record.name}</span>);
          },
        },
        {
          start_time: 1499730800,
          end_time: 1499739400,
          name: '小刘',
          render: (cell, record) => {
            return (<span>{record.name}</span>);
          },
        },
        {
          start_time: 1499835400,
          end_time: 1499849500,
          name: '小谢',
          render: (cell, record) => {
            return (<span>{record.name}</span>);
          },
        },
        {
          start_time: 1499930000,
          end_time: 1499937000,
          name: '小航',
          render: (cell, record) => {
            return (<span>{record.name}</span>);
          },
        },
        {
          start_time: 1500025000,
          end_time: 1500030000,
          name: '小张',
          render: (cell, record) => {
            return (<span>{record.name}</span>);
          },
        },
        {
          start_time: 1500035000,
          end_time: 1500082000,
          name: '跨时间的老王',
          render: (cell, record) => {
            return (<span>{record.name}</span>);
          },
        },
      ],
    };
  }

  render() {
    const sheetProps = {
      value: this.state.value,
      events: this.state.events,
      onChange: (value) => {
        this.setState({
          value,
        });
      },
      rowHeight: 40,
      showLine: true,
      direction: 'normal',
    };
    return (<div>
      <div style={{ width: '100%', margin: '100px auto', textAlign: 'center' }}>
        <WeekSchedule {...sheetProps} />
      </div>
    </div>);
  }
}

ReactDOM.render(<Test />, document.getElementById('__react-content'));
